### DISCUSSION TITLE
Onboarding

### DISCUSSION THREAD

**Discussion by iancooper** [BrighterCommand/Brighter#2045] [General]
Hi @omelianlevkovyc,

Let's me know if you can see this, and the we can help you get started. I've attached a screen shot of our 'help wanted' section.

As an alternative, we need to work on our docs, and I am currently reviewing our samples. I am working on some WebAPI examples. I have one done  for EF Core and I am working on a Dapper version of it. To cover of the most popular .NET data access libs we could do with another example for WebAPI and Marten.

The advantage of that would be that you could learn how to use Brighter, if you don't already know, and you can use the existing WebAPI and EFCore project to provide some ideas about how to write it.

Thoughts?

Ian

<img width="1284" alt="Screenshot 2022-03-25 at 15 29 48" src="https://user-images.githubusercontent.com/45537/160151596-4dd85cce-e4e4-4672-abac-8ac5cf418dd4.png">

**Comment by omelianlevkovych** [comment]
Hi there everyone 👍 
I agree that It'll probably be better to start with documentation so I can get onboard and have some look on the Brighter.

**Reply by iancooper** [reply]
@omelianlevkovych 

Guessing you mean the sample code suggestion?

We have a samples folder, you can find what we have there.

Creating a new sample, similar to the WebAPI EFCore one but using Marten would be great. There will be a tricky bit, which is we don't have Outbox integration for Marten yet. Don't worry. Ignore Outbox integration for the first version of the sample, and then we can  work together on that.

Now... V9 is fairly new out, and things have changed, so our existing docs, in the Docs repo, are probably really out of date.

So as you work through trying to understand WebAPI EF Core and write the Marten version, calling out the broken docs, and helping us fix them would be enormously helpful. Getting the docs to reflect V9 is the big challenge for the next few months.

**Reply by iancooper** [reply]
@omelianlevkovych Maybe as a suggestion:

* Can you get the WebAPI and EF Core example running?

Give it a try and raise any problems. We can then document in the readme for other folks. Because they will struggle just like you. Once we have you up and running with that example, we can look at modifying it for Marten.

**Reply by iancooper** [reply]
(You'll want get RMQ running using Docker as part of that, and that is a good step as it means you can use messaging middleware, which is sort of essential here)

**Reply by omelianlevkovych** [reply]
@iancooper got it now. Adding new sample seems achievable.
I will start with getting WebAPI and EFCore samples up and running, as you already suggested.
By the way, is experience with Marten required? And I assume that by Marten you mean this: https://martendb.io/, am I correct here?

**Reply by iancooper** [reply]
No experience needed (I have none). Hopefully, useful learning for you. As I say, skip the outbox on v1, or use in-memory. Then we can write the integration with outbox with you.

**Reply by omelianlevkovych** [reply]
@iancooper help needed. Seems I am trying to do something completely wrong here.
Is it possible that we miss GreetingsAdapters directory in the WebAPI_EFCore? I can't run docker compose unfortunatelly.
`ERROR: build path C:\personal\oss\samples\WebAPI_EFCore\GreetingsAdapters either does not exist, is not accessible, or is not a valid URL.`

What is the right way to build the samples? Should I use build.sh script, or it is okay to build it from VS?
By the way, when trying to build the solution in Visual Studio I am getting the following error message:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	NU1104	Unable to find project '.\Paramore.Brighter.ServiceActivator.Extensions.HealthChecks\Paramore.Brighter.ServiceActivator.Extensions.HealthChecks.csproj'. Check that the project reference is valid and that the project file exists.	GreetingsWorker	.\samples\ASBTaskQueue\GreetingsWorker\GreetingsWorker.csproj	1	
```
However, after unloading that example project, I was able to build succesfully and also run the WebAPI EFCore example (not all endpoints are working but I assume it is due to not having RMQ up and running).

And the last question, sorry if it is too silly. How can I run tests for WebAPI EFCore example project. I see that we have test.http there. Any hints?
Thanks in advance :)

**Reply by iancooper** [reply]
So, the project is not built to be run in docker compose (the issue you are seeing is probably to do with a docker compose file needing to be at the root) but stand-alone. We tend to use docker compose to provide RMQ or MySQL. There are docker compose files in the root of the Brighter project that you can use for that.

**Reply by iancooper** [reply]
In principle it should build and run from the command line, but I use Rider and tend to run it from the launchsettings.json file, which will give me the correct environment variables for dev, production etc https://github.com/BrighterCommand/Brighter/blob/master/samples/WebAPI_EFCore/GreetingsWeb/Properties/launchSettings.json That chooses between Sqlite and MySQL as well, you will want a MySQL instance (hence the docker files in the root of Brighter) if you choose production.

**Comment by iancooper** [comment]
BTW if you don't know RMQ, localhost:15672/ or localhost:55672/ will give you the management console (usr:guest pwd:guest) where you can see what is happening, which really helps

**Comment by iancooper** [comment]
RMQ will drop a message, if it has no consumer. So you have to spin up the consumer - before you send messages, or they will be lost. Doing a POST of a greeting to a user should send a message

**Comment by iancooper** [comment]
Finally, yes, the tests.http file will work with Rider or RESTClient in Code to let you run tests against the WebAPI that should then drive messaging

**Comment by iancooper** [comment]
Let me know how it all goes!!

**Comment by omelianlevkovych** [comment]
Thanks a lot. Will try it out 👍

**Reply by omelianlevkovych** [reply]
Short update. Currently I am a bit stack with setting up RMQ. 
It seems like I am doing something wrong when trying to connect publisher (GreetingsWeb) to my RMQ instance.
Here is the general message:`
RabbitMQ.Client.Exceptions.BrokerUnreachableException: 'None of the specified endpoints were reachable'`.
If more detailed logs are needed, please let me know.
Hoping to manage this on my own, but I decided that this worth sharing, so I won't spend too much on something silly and can invest that time a bit smarter.

Additional information (hoping relevant):
* RMQ Management works fine on http://localhost:15672/#
* `rabbitmq-diagnostics status` shows (in docker container):
```
Totals

Connection count: 0
Queue count: 1
Virtual host count: 1

Listeners

Interface: [::], port: 25672, protocol: clustering, purpose: inter-node and CLI tool communication
Interface: [::], port: 5672, protocol: amqp, purpose: AMQP 0-9-1 and AMQP 1.0
Interface: [::], port: 15672, protocol: http, purpose: HTTP API
```
* I haven't changed any source code, epsecially configuration in Startup yet.
* I was able to run Consumer (SalutationAnalytics), created a queue.
![image](https://user-images.githubusercontent.com/22483733/160259473-06f99bc2-445b-4526-a69f-9cd19dbb1b29.png)
* I do use the same creds: `guest; guest`
* my docker-compose.yml:
```
version: '3'

services:
  mysql:
    hostname: greetings_db
    image: mysql
    ports:
      - "3306:3306"
    security_opt:
      - seccomp:unconfined
    volumes:
      - my-db:/var/lib/mysql
    environment:
      MYSQL_ROOT_PASSWORD: "root"
    healthcheck:
      test: mysqladmin ping -h localhost -p$$MYSQL_ROOT_PASSWORD && test '0' -eq $$(ps aux | awk '{print $$11}' | grep -c -e '^mysql$$')
  rabbitmq:
    image: brightercommand/rabbitmq:3.8-management-delay
    ports:
      - "5672:5672"
      - "15672:15672"
    volumes:
      - rabbitmq-home:/var/lib/rabbitmq
volumes:
    rabbitmq-home:
      driver: local
    my-db:
      driver: local
```
I am planning to read more RMQ docs tomorrow as well; especially : [rabbit-mq-networking-troubleshooting](https://www.rabbitmq.com/troubleshooting-networking.html#inspecting-connections). 
It anyway will be very useful in the nearest future, and I am afraid that I am missing something very basic here.

**Reply by omelianlevkovych** [reply]
Finally, I have noticed that for production we do not use localhost, changed AMPQ uri to `amqp://guest:guest@localhost:5672` in the Startup and now I can see the connection in RMQ Managment.
Also RMQ now process messages:
![image](https://user-images.githubusercontent.com/22483733/160271953-781cdf95-7532-4350-8e9f-a90425e22d3c.png)
I am also able to send greetings over postman.

Now I am planning to spin up the consumer and debug the message receiving process; and than I will move to see whether tests are working.

**Reply by omelianlevkovych** [reply]
I have added phpmyadmin and connected to the mysql instance `hostname=greetings_db, pass&username=root`
I can see in the Salutations table the sended message:
![image](https://user-images.githubusercontent.com/22483733/160274006-5685fbdf-aacf-41fb-b564-64dd6d485e6b.png)
However, the Inbox table in the Salutations db is empty. Also I am not quite sure what I should see in the RMQ Managment:
![image](https://user-images.githubusercontent.com/22483733/160274037-f6d8677b-6b17-4c72-8e0d-ff5124d67656.png)

Shortly, to make it works (publisher-rmq-consumer) I have changed host to localhost in production; not sure it is the right approach.
Now I will try to make tests to pass.

**Comment by iancooper** [comment]
Hi, sounds like you have made some progress. Yes, checking the AMQP connection string is always a good idea, as you need to make sure it points to the instance you are using for testing (in this case localhost).

**Comment by iancooper** [comment]
You would expect to see the following:

* Message in Outbox for the outgoing message
* A historical record of the message via message rates
    * If you want to, you can stop the consumer, run the producer, see the message sitting in the queue, and then run the consumer to eat it; this lets you see the asynchronous nature of the workflow
* The inbox should have a copy, though you need to make sure that the inbox table exists in the db (sqlite or mysql depending on what you are choosing)

One thing to watch for with sqlite - its read-only for subsequent runs and fails, as only the creating process has rights to write by default, unless you explicitly grant those rights. It's a PITA for multiple-runs. If that is the case, just ensure you are using your MySql instance for the inbox too.

If you turn diagnostics up enough, you should get error messages about failed writes and why.

**Comment by iancooper** [comment]
Let me know if that helps at all.

**Comment by iancooper** [comment]
But fantastic to see that you have gotten so far.

**Comment by omelianlevkovych** [comment]
@iancooper yea, you are very helpful, I appreciate it a lot.
Also hoping I am not very annoying :sweat_smile:
* I do see messages in the Outpox table (in **Greetings** db); they also are getting dispatched
* I have stopped consumer and can now see that messages are staying in the queue. Nice.
Unfortunatelly, despite having the Inbox and Outbox tables in the **Salutations** db, I still missing data there. Only the Salutations table is getting populated after consumer get messages from the queue.
![image](https://user-images.githubusercontent.com/22483733/160281476-145c0136-9409-460a-a08f-ea6f5bd9ed11.png)
I am using Production launch settings and **MySQL** therefore, any ideas why my Inbox table is empty?

However, everything else seems to work. Lets see whether I can figure out how to run tests now. :wink:

**Reply by omelianlevkovych** [reply]
Oh okay, so tests.http its more like postman collection rather than some integrational tests.
@iancooper In case I will figure out why I don't have anything in Inbox table, what is our next step?
**updated**: I have tested with the endpoints in tests.http via VS Code extension, all seems to work.
The last question is whether I should have any data in the Inbox table (Salutations db).

**Reply by iancooper** [reply]
You are not annoying at all. Happy to onboard new people, and I learn a lot from figuring out how easy it is for new folks to get end-to-end.

What would be really valuable is if you can think of how we could update the readme.md for this sample, to make it easier. What did you not understand? Can you submit a PR with an update for the sample's readme.md for that? What you don't now is what most won't know, and is hard for us deep in the project to understand.

If the http calls all work there ought to be data in the inbox. If not, we might have a bug. There is an explicit attribute on the handler for the Inbox, but it is commented out IIRC. Try commenting it in, and see if that makes it work. In that case there may be an issue with the"auto" Inbox

**Reply by iancooper** [reply]
Once you can figure out how to get Brighter working, the steps would be:

- PR the sample readme if you can, to record what you had to discover
- Figure out if the Inbox has a bug

**Comment by iancooper** [comment]
After that, it's either, 

- We figure out what is wrong with the Inbox, or the sample
- We look at how we could replace EF Core with Marten (https://martendb.io/introduction.html) for another example
- - This will have a simple part (get it working without Outbox support) and a hard part (Outbox support for Marten)

**Reply by iancooper** [reply]
PS Yes, tests.http is just a lightweight version of postman effectively

**Comment by omelianlevkovych** [comment]
@iancooper I uncommented the attribute and inbox works now 😎  
I would appreciate to make the PR for this particular example update. 
However, I am not sure whether I have to understand how all that works, because honestly I don't.
So I see two possible ways of doing it, depends on what we need. 
* I can update readme with things I struggled when spinning it up (publisher, rmq&mysql, consumer). And than learn more by implementing Marten example (also writing doc for that, and probably I can than add more things I learned into doc for EFCore_MySQL). 
* Or I should basically spend more time figuring out what exactly Brighter does, all that clean architecture benefits it brings, etc.

**Reply by omelianlevkovych** [reply]
Minor note, I was walking through https://www.goparamore.io/ and I can see that some links are broken.

**Reply by iancooper** [reply]
>I uncommented the attribute and inbox works now
Feel free to raise a bug for that one, and I'll pick it up, not sure why it is failing

>I can update readme with things I struggled when spinning it up (publisher, rmq&mysql, consumer).
Feel free to do that, we can work on the PR with you, so don't worry too much about whether its "right" first time. The feedback on where our docs were lacking was the most useful.

>And than learn more by implementing Marten example
Try that but don't use an Outbox when you do. The Outbox for Marten will be more complex. So just figure out how to get a version of the example working, but with Marten instead. That should help you check that you can "use" Brighter. Once we have that running (and feel free to tweak the example if you want to, to learn more), we can talk about how Marten might work with an Outbox. Use the DocumentDb approach to Marten.

**Reply by omelianlevkovych** [reply]
> Feel free to raise a bug for that one, and I'll pick it up, not sure why it is failing

Sure, I will try to look into code again so I can maybe provide something meaningful as an issue. Pitty of me, but I only now have found out what actually the Outbox stands for. I should have read the paramore documetation much earlier.

> Feel free to do that, we can work on the PR with you, so don't worry too much about whether its "right" first time. The feedback on where our docs were lacking was the most useful.

I will try to make PR today/tomorrow.  
By the way I still have a question considering the docker compose.
> So, the project is not built to be run in docker compose (the issue you are seeing is probably to do with a docker compose file needing to be at the root) but stand-alone. We tend to use docker compose to provide RMQ or MySQL. There are docker compose files in the root of the Brighter project that you can use for that.

So if we have docker compose files in the root (so we can spin up containers like RMQ), why do we need another one in the example folder?

> So just figure out how to get a version of the example working, but with Marten instead. That should help you check that you can "use" Brighter. Once we have that running (and feel free to tweak the example if you want to, to learn more), we can talk about how Marten might work with an Outbox. Use the DocumentDb approach to Marten.

Got it sir. Thats seems to be an interesting :boom: task for me.

**Reply by omelianlevkovych** [reply]
jfi, I have created two issues
* one for the potential bug with Inbox
* another one is for the readme update for EFCore_WebApi example; It seems to be the right way before creating PR; in case I understand Contributing documentation correctly. Hoping to create PR soon.

also created the PR for the doc update

**Reply by omelianlevkovych** [reply]
> So if we have docker compose files in the root (so we can spin up containers like RMQ), why do we need another one in the example folder?

Also asked this in the PR, maybe we should continue discussion about docker compose there, idk

**Reply by preardon** [reply]
@omelianlevkovych let me know if I can be of any assistance

**Reply by omelianlevkovych** [reply]
@preardon Sure. 
Could you please have a look to my PR https://github.com/BrighterCommand/Brighter/pull/2054. For some reason CI for AWS has failed and I do not have permissons to rerun it.
Also I am not sure whether I should link the issues I created for the doc with the PR.
Thanks a lot in advance.

**Reply by preardon** [reply]
@omelianlevkovych This is a problem with our AWS Test being flaky, I will have a look at the PR and add any appropriate comments

**Comment by omelianlevkovych** [comment]
@preardon Okay, thanks for your comment, it motivates a lot. 
By the way I have pushed hopefully the last updates to this PR.

I will start with Marten example today, so probably will have some questions sooner or later.

**Reply by iancooper** [reply]
I merged the PR for you.

**Reply by iancooper** [reply]
Sorry it took so long, busy day.

PS @preardon is a good source of help if I am not around (along with @holytshirt and @jonnyollifflee)

**Comment by omelianlevkovych** [comment]
@iancooper , @preardon  Hi guys, do we have some documentation about how to properly configure the Brighter in Startup?
I am currently having an issue with DI that It doesn't see Handler for the Command.
`System.ArgumentException: No command handler was found for the typeof command GreetingsPorts.Requests.AddPerson - a command should have exactly one handler.`
I assume this extensions should figure out my DI setup AutoFromAssemblies.
I have pushed the code to `webapi_marten_example` branch

* Do I have to use Message Broker even though I am using CQRS (basically the publisher API; for instance, to create a Person)
* Does Brighter support Outbox pattern under the hood, can I somehow setup distributed transacrtion? And for my case `UseInMemoryOutbox` should be just enough so far?

**Reply by preardon** [reply]
@omelianlevkovych  I will have a look at this now. out of interest why did you do this in dotnet 5 instead of 6?

**Reply by omelianlevkovych** [reply]
@preardon hah, thats good question. I have actually started doing it in dotnet 6. However, than I realised that personally for me it'll be a bit easier to use version 5. I will not have to think for now whether to create some extensions or have all the host builder related logic in the Program class; also the nullable references, etc.
But yea, I think it will be a good idea to move it to dotnet 6 as far as I will get a bit familiar with Brighter.

**Reply by preardon** [reply]
The main reason I ask is because I'm trying to debug this on an m1 mac and I haven't configured Rider to work correctly with dotnet 5 so I'm dry reading 😆

**Reply by omelianlevkovych** [reply]
@preardon Hah, sorry for that.
I can update to dotnet 6 version, rebuild and commit. It than should be easier for you I guess. Should I do it now?

**Reply by preardon** [reply]
<img width="1119" alt="image" src="https://user-images.githubusercontent.com/7712876/161398220-ca1275ad-c70f-4073-a75d-3da0f0574bce.png">
Got it working

**Reply by preardon** [reply]
There are a few things that I needed to do here

Firstly it appears that it wasn't running from the correct dll.

I first tried being a little more specific
`.AsyncHandlersFromAssemblies(typeof(AddPersonHandlerAsync).Assembly)`

This picked up your other handler that had an unmet dependency so I added

`services.AddSingleton<GreetingsEntityGateway>();`

This then started complaining about too many handlers, so I removed the AutoFromAssemblies and it start working.

to confirm my suspicion of Lazy Loading I removed the AsyncHandlersFromAssembly as above and added

`var tst = new AddPerson("Paul");`

and it continued to work!!

tldr; lazy loading stopped Auto from working, if you were to load a service earlier then the dll should be loaded all you need to add is
`services.AddSingleton<GreetingsEntityGateway>();`

I am not sure of your knowledge level, please let me know if you would like any more clarification than above

**Reply by omelianlevkovych** [reply]
Thanks a lot, that is very clear explanation. Obviously it also works for me now 👍

**Reply by preardon** [reply]
Excellent, let me know if I can be of any more help

**Comment by omelianlevkovych** [comment]
Do you guys aware about any sqlite similar document database, which I can use for the Development environment for the Marten example.

Or for now its ok to use postgresql for both.
[marten doc about dev and prod envs](https://martendb.io/configuration/optimized_artifact_workflow.html)

**Reply by omelianlevkovych** [reply]
And also I am wondering whether I should create a PR.
I have added the Greetings + Marten part only for now.
It still lacks the Salutations (consumer), RMQ config and outbox pattern, update to net6.

It would be nice to get some critique earlier on, before I made something really stupid and than had to rollback it all.
On the other hand, I don't want to bother much.

**Reply by jonnyolliff-lee** [reply]
@omelianlevkovych 

> Do you guys aware about any sqlite similar document database, which I can use for the Development environment for the Marten example.

I'm not up on Marten, but I'll try and do some reading and see. Someone else might have a better idea.

> And also I am wondering whether I should create a PR.

it's worth creating a "Draft" PR if you want to get some early thoughts, but you're not completely ready.

![image](https://user-images.githubusercontent.com/101091884/162379814-a4d1873c-b560-421d-b2ed-46c583f93f0d.png)

**Reply by omelianlevkovych** [reply]
@jonnyolliff-lee thanks a lot!

**Comment by omelianlevkovych** [comment]
@iancooper Hey, Is the example for Marten still a thing? I just can not find the issue anymore.
I would like to get back to that and finalize the PR (actually I committed from the wrong git user and will probably recreate the PR).

**Comment by iancooper** [comment]
@omelianlevkovych Your PR still exists: https://github.com/BrighterCommand/Brighter/pull/2076 but you may need to merge with master to pick up recent changes. Welcome back!
